XML macros: add named yields, tokenized macros, tokens for attributes, and fix replacement of toplevel yield

bernt-matthias opened the pull request:
Originally I wanted to find out what exactly happens if there are multiple yield tags.. hoping that I can somehow control that each yield is replaced with the corresponding child of the expand tag. This did not work. So I added named yields and two other extensions that I were missing in the past:

**Named yields**

Essentially `<yield name="XYZ"/>` is replaced by the content of `<template name="XYZ"/>`. Unnamed yields work as before and named and unnamed yields can be mixed.

An example:

```xml
    <xml name="named_yields_example">
      <conditional>
        <param type="select">
          <option value="a">A</option>
          <option value="b">B</option>
          <yield name="more_options"/>
        </param>
        <when value="a">
          <param type="select">
            <yield />
          </param>
        </when>
        <when value="b">
          <param ... />
        </when>
        <yield name="more_whens">
      </conditional>
    </xml>
```

```xml
        <expand macro="named_yields_example">
          <token name="more_options">
            <option value="c">C</option>
          </token>
          <token name="more_whens">
            <when value="c">
              <param type="select">
                <yield />
              </param>
            </when>
          </token>
          <options from_data_table="tophat2_indexes">
            <filter type="sort_by" column="2"/>
            <validator type="no_options" message="No genomes are available for the selected input dataset"/>
          </options>
        </expand>
```

The choice of `token` as tag name may be sub-optimal... since tokens are also used to parametrize macro parameters. Should be easy to change if desired. 


**Allow that the name of nested macros can be controlled by tokens:**

For instance 

```xml
<xml name="example">
    <expand macro="@TOKEN_NAME@"/>`
</xml>
```

See https://github.com/galaxyproject/galaxy/blob/372439c088ee6b7b7b1a5751713891582928d39e/test/unit/tool_util/test_tool_loader.py#L611 for a complete example

**Allow token expansion also for attributes**

```xml
<tool>
    <macros>
        <token name="__ATTR_NAME">name</token>
    </macros>
    <another>
        <tag __ATTR_NAME="blah" />
    </another>
</tool>
```

Here the token can't be surrounded by `@` character (i.e. `@ATTR_NAME@`) since the xml file needs to proper xml (because it is parsed before token/macro expansion). But it seems that for tokens surrounding the token by `@` is a (undocumented) convention only, anyway. 

Docs have been added here: https://github.com/galaxyproject/planemo/pull/1212 (not all changes are yet documented, but will be when this is merged).

I'm happy if there are more suggestions to make this more expressive / powerful / elegant :). Real recursion would be cool but also dangerous .. also could not think of a way to express a stopping condition.


Fix: If there are multiple top level yield statements only the last was replaced

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
fix replacement of toplevel yield

if there are multiple top level yield statements
only the last was replaced

bernt-matthias committed:
remove apparently unnecessart code

bernt-matthias committed:
even simpler _expand_yield_statements

bernt-matthias committed:
add docstring

bernt-matthias committed:
XmlMacroDef: store the macro element

instead of the list of children

mv parsing of token_quote out of the for loop, since otherwise
tokens that are parsed before token_quote will get the default
and only the ones after will get the desired token_quote

bernt-matthias committed:
fix unit test

use galaxy.util.xml_to_string for export since lxml is not always
available?

bernt-matthias committed:
implement named yields

yield tags with a name attribute will be replaced

```
<xml>...<yield name="xyz"/>...</xml>
```

by the content of the token with the corresponding name:

```
<expand>...<token name="xyz">CONTENT</token>...</expand>
```

The token must be direct child of the expand tag

like for unnamed yields all yields with the same name will
be replaced by the token's content.

Named yields are replaced befor unnamed yields.

When replacing unnamed yields tokens are ignored.

bernt-matthias committed:
add test for the limited recursion possibilities

bernt-matthias committed:
use more xpath in test_tool_loader

got myself used to the unit test and thought it might be nice
to use more xpath

undo a unnecessary change in xml_macros.py

bernt-matthias committed:
move doctests to unit tests

bernt-matthias committed:
some more extensions for xml macros

1. allow that the name of nested macros can be controlled by tokens:
eg `<expand macro="@TOKEN_NAME@"/>`

therefore:

- in load_with_references childs of the macros element are
  temporarily removed. otherwise the replacement of the nested macros
  also happens for the expand tags in macros and it seems really
  difficult to iterate a tree excepting a subtree
- in _expand_macro replace tokens first and then consider
  nested macros

see test test_loader_specify_nested_macro_by_token

2. allow token expansion also for attributes (before only attribute
values)

this only works if the token is valid xml i.e. they can't be surrounded
by `@`

see test test_loader_token_attribute_name

3. split unit tests in separate functions

and add a few (e.g. nested tokens ...)

bernt-matthias committed:
fix tool loader if no macros are present

and add unit test for this case

bernt-matthias committed:
check for cycles in nested macros

mvdbeek commented on lib/galaxy/util/xml_macros.py:
```diff
@@ -17,13 +17,30 @@ def load_with_references(path):
+    # that are not included in the macros node
+    macros_el = _macros_el(root)
+    if macros_el is not None:
+        macros_copy = deepcopy(macros_el)
```
deepcopy essentially pickles and unpickles the object, this could be detrimental to startup speed. Would a `.copy()` be sufficient ?

bernt-matthias replied:
True. I remember that if I do not deepcopy then the following clear also removes the copied elements. Maybe calling `remove(subelement)` for each child helps. 

In the end this is only a workaround: I was not able to find a way to iterate only the `expand` elements that are not included in the `macros` subtree. With lxml it might be possible, but it appears to me that the code needs to be compatible with xml from the stdlib and lxml, or am I wrong?

mvdbeek replied:
You could import `LXML_AVAILABLE` and make that conditional on this (if it works).

mvdbeek commented on lib/galaxy/util/xml_macros.py:
```diff
@@ -134,48 +164,60 @@ def _expand_macros(elements, macros, tokens):
     assert macro_name is not None, "Attempted to expand macro with no 'macro' attribute defined."
+
+    # check for cycles in the nested macro expansion
+    assert macro_name not in visited, f"Cycle in nested macros: already expanded {visited} can't expand '{macro_name}' again"
```
Are you sure this must be a cycle ? Also list lookups can be expensive, maybe make visited a set or an ordered set?

bernt-matthias replied:
> Are you sure this must be a cycle ? 

Quite. Essentially this is a DFS of the expansions and if a previously expanded macro is expanded again we have a cycle. Note that visisted is kind of a stack (i.e. the macro names are removed at the end of the function.)

> Also list lookups can be expensive, maybe make visited a set or an ordered set?

Then I will go for ordered set, since otherwise the test can not check for the message (or I just test for the prefix `"Cycle in nested macros"`.

mvdbeek commented on lib/galaxy/util/xml_macros.py:
```diff
@@ -134,48 +164,60 @@ def _expand_macros(elements, macros, tokens):
                 break
-            _expand_macro(element, expand_el, macros, tokens)
+            if visited is None:
+                v = list()
```
```suggestion
                v = []
```

mvdbeek reviewed (commented):
We should run this against the iuc as well.

bernt-matthias committed:
Update lib/galaxy/util/xml_macros.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

bernt-matthias committed:
use copy instead of deepcopy

mvdbeek reviewed (approved):
This looks good, but let's have a look at the IUC tests in https://github.com/galaxyproject/tools-iuc/issues/4332

bernt-matthias committed:
fix deprecation warning

if macros copy is actually an `Element` then this gives a deprecation
warning:

xml_macros.py:41: FutureWarning: The behavior of this method will change in future versions. Use specific 'len(elem)' or 'elem is not None' test instead.
  if macros_copy:
